### DISCUSSION TITLE
deprecate mysqlclient and support only PyMySQL

### DISCUSSION THREAD

**Discussion by markuman** [ansible-collections/ansible.mysql#645] [Polls]
@Akasurde @Andersson007 @Aohzan @FlorianPerrot @Hramoff @Jorge-Rodriguez  @NicolasGoeddel @TimWolla @asarubbo @betanummeric @chriscroome @cwchristerw @danbuntu @dennisurtubia @drzraf @ehudkaldor @eryx12o45 @francescsanjuanmrf @fs30000 @gotmax23 @kmarse @laurent-indermuehle @lkthomas @mario-pranjic @n-cc @rafi0101 @rajcspsg @razerrazer @resuni @robertsilen @sascha-andres @sczelo @tangel0v @tompal3 @wfelipew @yasyx @yosefy @rsicart @fhamme @steveteahan @bmildren @koleo @GhostLyrics @ziegenberg @hubiongithub @eRadical @webknjaz @aminvakil @MrMegaNova @aneustroev @felixfontein 

There are considerations to drop support for `mysqlclient` driver in `communiy.mysql` and move forward only with `PyMySQL` driver.

I've mentioned you here because you've contributed in the near past to `community.mysql` in some form (_issue and code commits_), so you've identified yourself as an active user of this collection.

##### edit 1

let's keep this poll open for at least one week.

##### edit 2

dropping means that `mysqlclient` will be still supported for several months.  
It means that is must be marked as deprecated in the changelog of a major release and one major release later, it will be dropped.

**Comment by cwchristerw** [comment]
I use only pymysql and mysqldump.

**Comment by chriscroome** [comment]
Am I correct in understanding that mysqlclient support is only needed for servers that don't / can't have PyMySQL installed? If that is the case I'd be happy to see the dropping of mysqlclient support.

**Reply by markuman** [reply]
Yes.  
First, it tries to use `PyMySQL`.  https://github.com/ansible-collections/community.mysql/blob/main/plugins/module_utils/mysql.py#L22  

If it's not avaiable, it tries `mysqlclient`.

**Comment by NicolasGoeddel** [comment]
I had issues with mysqlclient because it always tried to connect to a socket when when I set the host to `localhost.` when instead I wanted it to connect to a local HAProxy and using `localhost` as the CN for checking against the self-signed SSL-certificate on my Galera cluster.

With PyMySQL it worked fine. So I am fine with dropping the support of `mysqlclient` but in fact I also don't care.

**Reply by GhostLyrics** [reply]
Technically that's MySQL detail because `localhost` makes it prefer sockets while `127.0.0.1` is what you'd have used if it wasn't for your self-signed SSL-cert check.

**Reply by felixfontein** [reply]
I'm not sure anymore, since I switched from mysqlclient to PyMySQL a long time ago, but I also remember that I had trouble with mysqlclient. Though no idea what specifically :) So :+1: for dropping mysqlclient and keeping only PyMySQL.

**Comment by sascha-andres** [comment]
I have to manage some older versions of distributions where Python is especially outdated. This leads to running Ansible partially from a docker container which is a pain. Relying on a Python library (no objections specifically) would add another pain point when it comes to updates on those machines.

From an Ansible support POV this is totally reasonable though

**Reply by markuman** [reply]
I feel your pain. I have to maintain OEL 7 and 8, which are official not supported anymore by ansible (_`ansible-core` > 2.16 drops support for python 3.6_). I'm forced to maintain venv for those targets on our control nodes (_yes, docker container would also work here ..._).  

To be clear. Dropping means, that it first will be announced in the changelog with the next major release.  
And another major release later, it will be dropped. ... _something like that_.  
So support for `mysqlclient` will remain for several months....maybe one year?

**Reply by laurent-indermuehle** [reply]
@sascha-andres : Isn't mysqlclient also a "python" library?

**Reply by cwchristerw** [reply]
You're correct @laurent-indermuehle

**Reply by felixfontein** [reply]
In fact PyMySQL is a **pure** Python implementation, while mysqlclient is a Python library with a C part (IIRC) that needs to be compiled. Which generally makes it harder to install mysqlclient (in particular if wheels aren't available for your system).

I guess one big question is how outdated your Python is, and whether the latest versions of PyMySQL that support that Python are still supported by community.mysql.

**Reply by GhostLyrics** [reply]
> In fact PyMySQL is a pure Python implementation, while mysqlclient is a Python library with a C part (IIRC) that needs to be compiled. Which generally makes it harder to install mysqlclient [...]

That's exactly why my team switched to PyMySQL.

**Comment by laurent-indermuehle** [comment]
It would be great if the 5% who chose "keep mysqlclient support" could share their reasons. If we decide to drop support for this connector, would switching to PyMySQL be feasible for you, or are there any obstacles?

**Comment by markuman** [comment]
I'm closing this pull. the result is clear.
